### ISSUE TITLE
[Logistics] Some questions/suggestions/opinions about the grading scheme

### ISSUE THREAD

**Issue by KunJeong** [kaist-cp/cs420#177]
I've waited a while to talk about this, and since all the homework and final have been announced, I think it is the right time to do so. First of all, I would like to apologize if any of the words below feel rude / inappropriate. I have tried my best, but I found it quite difficult to state my opinion clearly without using phrases that sound either whiny or judgmental.

1. Is the grading policy for homework actually binary(0 or full score)?
I initially didn't really think it would be this way. _"Maybe HW1 is too easy, hence the binary grading" "Maybe HW2 is less..important..?"_ However, we've come this far, and all the homework are in fact binary! I wanted to make sure, since you mentioned the grading scheme only once in the lectures.
I understand the reason for this as I also believe that perfect/not perfect is all that matters to a programmer. However, I'm quite unsure if it is a good policy in terms of measuring the students' effort. TBH, I haven't seen such a policy in my short time as a KAIST CS major. And while I also agree that other methods, like percentage of test cases passed, are not perfect as they do not accurately reflect the amount of effort that was put in, a binary grade actually feels a lot worse in "reflecting the amount of effort", especially since some homework (like HW2) are quite massive.
Since the grading policy is binary, students cannot "choose" to deny certain parts of the homework, like for instance stop working on some homework after reaching 90%, which happens pretty often for other CS courses. One could see this as a good thing, but this is **already a very demanding course**, and we also have duties other than this course. Also, if everyone is forced to go through these homework without a choice, and everyone takes the time and effort to do hw1-7 perfectly, is everyone going to get a good to perfect grade? Probably not, since it's probably relative grading and our performance for final can't be all the same. This is where the second question occurs.

2. Would I get a 0 on the final if one of my homework is not perfect?
The fact that the final project, which is 20% of the score, actually **_has all of the previous homework as a "dependency"_** bothers me a lot. Would I get a 0 if I fail to finish one of the homework? If so, it would kind of force the students to do every inch of every homework perfectly, which brings us back to something similar to the first issue: we're forced to put in a massive amount of effort, but we're not actually graded based on our efforts. (I have no idea about the actual grades, so I'm speaking in general terms)

3. (Some details to support my opinion)
Following the GitHub repo, I can see that a lot of the students are either still stuck on HW2, or have recently finished HW2 and are moving on to HW3 and 4, even though roughly 3/4 of the semester has passed. It is hard to know for sure, but we're supposed to follow the pace of the lecture, meaning we should be done with HW6 by now, which seems impossible to me. Also the request to share each other's progress was rejected (#73).
Some statistics to support this - at the moment, there are the following number of issues for each homework (may be inaccurate):
HW1: 11, HW2: 37, HW3: 5, HW4: 4, HW5: 0, HW6: 1, HW7: 0
One could argue that hw2 is the most difficult, but I think you get the point. I'm pretty sure I'm not the only one having trouble following the pace of the lectures/homeworks. The load of this course is otherworldly, and I've taken some pretty demanding courses including OS. The details of the final project at this point is something that fills me with despair.

Of course, one opinion is far from enough to change a grading policy. But since we are doing online courses and there is very limited chances to communicate real-time, it is difficult to see how others are doing and how others feel about these things. I'm not trying to start a revolution or something, I just felt this way and **wanted to know if there were any others who felt the same**, I thought that **to know others' opinions I should do something first**, and finally I wanted to **let you know** since it is your first semester of instructing Compiler Design.

Thank you for reading this very long post.

**Comment by jeehoonkang** [comment]
Thank you for raising up the issue.  Before answering my opinion and decision, I'd like to hear from other students as well.

**Comment by pr0cf5** [comment]
I partially agree with the opinion above, but not fully. The following are my ideas.
First, I agree that for each homework the score must be proportional to the number of testcases passed instead of being all or none. As with the fuzzer, I think that credit should be given only if all 10 runs are successful.

Second, I agree on the fact that Irgen and Asmgen has rather lots to implement and takes a lot of time. It took a lot of time for me to do Irgen. But there were also several actions taken that eased the burden of doing it. 

For example, the professor was more than kind to provide a lot of codes for Irgen. Also, lots of students voluntarily found bugs in existing tests, reported them and fixed quickly. There was no midterm exam and there will be no final exam. The amount of theoretical content is not overwhelming, as in other CS classes that I took. Also, the projects (in my opinion) are well organized and fun.

So in conclusion, I agree that grading should be incremental. But for other issues, I am unsure.

**Comment by Ngiong** [comment]
My opinion:
1. I totally agree that grading 1/0 for HW2 is pretty unfair, w.r.t. the efforts that we are putting into for HW2. TBH I'm still stuck at array translation in HW2 and still considering to skip/sacrifice this part because I still have lab works (I'm a graduate student), other courses to catch up, etc.. 
Also, I am totally scared when I passed examples provided in HW2 (which is already difficult for me), my implementation does not pass the fuzzer (which still I get 0 scores for this)

2. Yes, I also totally agree that a final exam that wraps up all of our homework might be impossible for me too, (but what can I do 😭) 
and I think HW7 is another type of HW2 (so I have 2x HW2-type homework remaining to solve).
Luckily for my case, I've worked on all optimizations (I hope they're correct) so I can focus on HW2 or HW7 until July, and I will be more than grateful if I can finish one of these 😭 😭 

3. Yes, I'm also surprised that the fact #173 is created 2 days ago, and was never asked previously, does that mean no one has worked on HW4?

But this is just my opinion. FYI, I'm still adapting to new KAIST life because this is my first semester 
(and unfortunately I take this Compiler Design course with a lot of homework, I'm really hoping the rest of my life in KAIST will be a lot easier than this course 😁 ), 

and my opinion might not be a good measurement since I might be biased with how courses are taught in my previous institution.

p.s.: Although I think this course's homework is very difficult for me, I'm still enjoying learning compiler & optimization techniques from you. Thank you.

**Comment by christofides** [comment]
Hello
1. I also think it seems reasonable to give partial points if you pass the static homework cases for homework 2. 

2. Perhaps it could be possible to get partial points on the final project if you can show that your optimizations works when using the model-solution ir-gen (assuming that homework 2 are not fully completed). 

3. I am done with hw4, hw 3-4 were a lot quicker/easier than hw2 for me. Althought I didnt see the error in the issue #173 so I will have to double check if I missed something :).

**Comment by jeehoonkang** [comment]
Here are my opinion and decision.

- Regarding binary grading scheme: I'd like to stick to this grading scheme because it's how things will be usually evaluated for the rest of your career.  I understand it may be bad for educational purposes, but I implemented a few policies that mitigate the problem:

    + All homework/project assignments are due on the 8th of July, which is probably the latest possible date for this semester.  And all assignments are announced at least a month before the due date.  So you're not required to implement a perfect code for a short period of time, which I believe is a little bit brutal.  I decided to let you write the right code for a long time, instead of writing some code for a short time.

    + The grader is (or will be) provided for all assignments, and the grading results will be immediately announced after you submit your code.  Further, you can submit codes multiple times.  So you'll not be suddenly surprised by zero points due to previously unknown reasons.  If you'll get zero points, you know the reasons and you have time to fix it.  I know it's not a usual grading scheme, but I believe it's more educational.

    + Still, many students think hw2 and hw7 are too big to be graded in a binary manner.  I understand that's a valid concern, so I decided to provide a "smaller grader" for both assignments.  The new "hw2-small" and "hw7-small" will ask to implement only a subset of the assignments.  So if you have a complete hw2 solution, you'll get double the points; if you only have a partial solution that is sufficient for hw2-small, you'll get the points for hw2-small.  As previously announced, all assignments have the same weight.  The exact grading criteria of hw2-small and hw7-small will be announced by the 8th of June.  (I don't want to change the grading criteria after one month before the due date.)

- Regarding final project: I'd like not to change the project.  The sole purpose of this course is to implement a C compiler, and there's nothing else adequate for you to implement as a final project.  And as described in the slides, you can think of it as a bonus project: if you'll do good at it, you'll get higher letter grades; if you won't, then nothing bad or good happens.  Just in case that the nuance is unclear, I'm basically saying that (1) there's no final exam; (2) I'll give **extra** credits for those who produce a competitive compiler; and (3) I'm **not penalizing** those who don't.

- **In short**: I'll create two new nominal assignments: hw2-small and hw7-small.  If you already finished those, you don't need to care much.  If not, please wait a few days for me to announce the scope of those small assignments.

**Comment by KunJeong** [comment]
Thank you very much. Personally I think the solution is perfect - you've addressed all the issues that I have raised without sacrificing any of the great things we can get from the course. I really appreciate the many things we can learn from this course, and I also really enjoy the lectures and the homework. And I want to thank you for the massive amount of skeleton code, and the help you give to the students via the issue tracker. I apologize again if the original post seemed to imply otherwise. Thank you once more for gathering and respecting the students' opinion.

**Comment by jeehoonkang** [comment]
@KunJeong thank you again for the feedback.  I'm learning a lot from the feedback and questions of the students in the making of this course.

I'm leaving this issue open until the follow-up measures will be implemented.

**Comment by jeehoonkang** [comment]
Continuing the discussion in #195
